Libraries/Opc.Ua.Client/Subscription.cs

```diff
@@ -1819,42 +1842,81 @@ private async Task<(bool, UInt32Collection, UInt32Collection)> GetMonitoredItems
-                m_publishTimer = new Timer(OnKeepAlive, keepAliveInterval, keepAliveInterval, keepAliveInterval);
+                Interlocked.Exchange(ref m_lastNotificationTime, DateTime.UtcNow.Ticks);
+                m_keepAliveInterval = (int)(Math.Min(m_currentPublishingInterval * (m_currentKeepAliveCount + 1), Int32.MaxValue));
+#if NET6_0_OR_GREATER
+                var publishTimer = new PeriodicTimer(TimeSpan.FromMilliseconds(m_keepAliveInterval));
+                Task.Run(() => OnKeepAliveAsync(publishTimer, CancellationToken.None));
+                m_publishTimer = publishTimer;
+#else
+                m_publishTimer = new Timer(OnKeepAlive, m_keepAliveInterval, m_keepAliveInterval, m_keepAliveInterval);
+#endif
 
                 if (m_messageWorkerTask == null || m_messageWorkerTask.IsCompleted)
                 {
-                    m_messageWorkerShutdownEvent.Reset();
-                    m_messageWorkerTask = Task.Run(() => PublishResponseMessageWorkerAsync());
+                    Utils.SilentDispose(m_messageWorkerCts);
+                    m_messageWorkerCts = new CancellationTokenSource();
+                    m_messageWorkerTask = Task.Run(() => PublishResponseMessageWorkerAsync(m_messageWorkerCts.Token));
                 }
             }
 
             // send initial publish.
-            m_session.BeginPublish(Math.Min(keepAliveInterval, Int32.MaxValue / 3) * 3);
+            m_session.BeginPublish(BeginPublishTimeout());
         }
 
+#if NET6_0_OR_GREATER
+        /// <summary>
+        /// Checks if a notification has arrived. Sends a publish if it has not.
+        /// </summary>
+        private async Task OnKeepAliveAsync(PeriodicTimer publishTimer, CancellationToken ct)
+        {
+            try
+            {
+                while (await publishTimer.WaitForNextTickAsync(ct).ConfigureAwait(false))
+                {
+                    if (!PublishingStopped)
+                    {
+                        continue;
+                    }
+
+                    HandleOnKeepAliveStopped();
+                }
+            }
+            catch (Exception ex)
+            {
+                Utils.LogError(ex, "");
+            }
+        }
+#else
         /// <summary>
         /// Checks if a notification has arrived. Sends a publish if it has not.
         /// </summary>
         private void OnKeepAlive(object state)
+        {
+            if (!PublishingStopped)
+            {
+                return;
+            }
+
+            HandleOnKeepAliveStopped();
+        }
+#endif
+
+        /// <summary>
+        /// Handles callback if publishing stopped. Sends a publish.
+        /// </summary>
+        private void HandleOnKeepAliveStopped()
         {
             // check if a publish has arrived.
             PublishStateChangedEventHandler callback = null;
 
             lock (m_cache)
```

Comment:
One could use a semaphore and await the lock here, then the periodic timer might make more sense.  Question is though how much lock contention there is on m_cache.  Probably not that much, or?.